File: src/python/titanic_pipeline.py
import os
import sys
import pandas as pd
import numpy as np
from pathlib import Path

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

DATA_DIR = Path(os.getenv("DATA_DIR", "src/data"))

def log(msg):
    print(f"[INFO] {msg}")

def load_csv(name):
    path = DATA_DIR / name
    if not path.exists():
        log(f"ERROR: {name} not found at {path.resolve()}.")
        sys.exit(1)
    log(f"Loading {name} from {path.resolve()}")
    return pd.read_csv(path)

def engineer_features(df: pd.DataFrame) -> pd.DataFrame:
    log("Engineering features...")
    out = df.copy()

    out["Title"] = out["Name"].str.extract(r",\s*([^\.]+)\.", expand=False)

    out["CabinFirstLetter"] = out["Cabin"].astype(str).str[0]
    out.loc[out["Cabin"].isna(), "CabinFirstLetter"] = np.nan

    out["FamilySize"] = out["SibSp"].fillna(0) + out["Parch"].fillna(0) + 1

    out["TicketGroupSize"] = out.groupby("Ticket")["Ticket"].transform("count")

    keep = [
        "Survived", "Pclass", "Sex", "Age", "SibSp", "Parch", "Fare",
        "Embarked", "Title", "CabinFirstLetter", "FamilySize", "TicketGroupSize"
    ]
    cols = [c for c in keep if c in out.columns]
    out = out[cols]

    log(f"Feature columns now: {list(out.columns)}")
    return out

def main():
    log(f"Expecting Titanic CSVs in: {DATA_DIR.resolve()}")
    train = load_csv("train.csv")
    test  = load_csv("test.csv")

    log(f"Train shape: {train.shape}")
    log(f"Test  shape: {test.shape}")

    log("Preview missingness (top 10 columns):")
    print(train.isna().sum().sort_values(ascending=False).head(10))

    train_fe = engineer_features(train)
    test_fe  = engineer_features(test)

    if "Survived" not in train_fe.columns:
        log("ERROR: 'Survived' column not found in train.csv.")
        sys.exit(1)

    y = train_fe["Survived"].astype(int)
    X = train_fe.drop(columns=["Survived"])

    numeric = ["Age", "SibSp", "Parch", "Fare", "FamilySize", "TicketGroupSize"]
    categorical = ["Pclass", "Sex", "Embarked", "Title", "CabinFirstLetter"]

    log(f"Numeric features: {numeric}")
    log(f"Categorical features: {categorical}")

    numeric_tf = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler())
    ])
    categorical_tf = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("ohe", OneHotEncoder(handle_unknown="ignore"))
    ])

    pre = ColumnTransformer(
        transformers=[
            ("num", numeric_tf, numeric),
            ("cat", categorical_tf, categorical)
        ]
    )

    clf = LogisticRegression(max_iter=200)

    pipe = Pipeline(steps=[
        ("pre", pre),
        ("clf", clf)
    ])

    log("Fitting Logistic Regression on full training set...")
    pipe.fit(X, y)

    y_pred = pipe.predict(X)
    acc = accuracy_score(y, y_pred)
    log(f"TRAIN accuracy (on full train.csv): {acc:.4f}")

    log("Generating predictions on test.csv...")
    test_preds = pipe.predict(test_fe)
    test_out = pd.DataFrame({
        "PassengerId": test["PassengerId"],
        "Survived": test_preds.astype(int)
    })
    out_path = Path("python_submission.csv")
    test_out.to_csv(out_path, index=False)

    log(f"Saved predictions to: {out_path.resolve()}")
    log("Note: test.csv has no 'Survived' column; skipping accuracy check as instructed.")

if __name__ == "__main__":
    main()

File: src/r/titanic_pipeline.R
library(readr)
library(dplyr)
library(caret)
library(stringr)
library(tidyr)

log_msg <- function(msg) cat(paste0("[INFO] ", msg, "\n"))

DATA_DIR <- "src/data"

log_msg(paste("Expecting Titanic CSVs in:", normalizePath(DATA_DIR)))
train <- read_csv(file.path(DATA_DIR, "train.csv"))
test  <- read_csv(file.path(DATA_DIR, "test.csv"))

log_msg(paste("Train shape:", paste(dim(train), collapse = " x ")))
log_msg(paste("Test  shape:", paste(dim(test), collapse = " x ")))

engineer_features <- function(df) {
  log_msg("Engineering features...")
  
  df <- df %>%
    mutate(
      Title = str_extract(Name, ",\\s*([^\\.]+)\\.") %>%
        str_replace_all(",\\s*|\\.", "") %>%
        str_trim(),
      CabinFirstLetter = ifelse(is.na(Cabin), NA, substr(Cabin, 1, 1)),
      FamilySize = coalesce(SibSp, 0) + coalesce(Parch, 0) + 1
    )
  
  keep <- c("Survived", "Pclass", "Sex", "Age", "SibSp", "Parch",
            "Fare", "Embarked", "Title", "CabinFirstLetter", "FamilySize")
  keep <- intersect(keep, names(df))
  df <- df[, keep]
  log_msg(paste("Feature columns now:", paste(keep, collapse = ", ")))
  return(df)
}

train_fe <- engineer_features(train)
test_fe  <- engineer_features(test)

if (!"Survived" %in% names(train_fe)) {
  stop("[ERROR] 'Survived' column not found in train.csv.")
}

train_fe$Sex <- as.factor(train_fe$Sex)
train_fe$Embarked <- as.factor(train_fe$Embarked)
train_fe$Title <- as.factor(train_fe$Title)
train_fe$CabinFirstLetter <- as.factor(train_fe$CabinFirstLetter)

log_msg("Fitting logistic regression on full training set...")
model <- glm(Survived ~ Pclass + Sex + Age + SibSp + Parch + Fare +
               Embarked + Title + CabinFirstLetter + FamilySize,
             data = train_fe, family = binomial())

pred_train <- ifelse(predict(model, type = "response") > 0.5, 1, 0)
train_acc <- mean(pred_train == train_fe$Survived)
log_msg(paste("TRAIN accuracy (on full train.csv):", round(train_acc, 4)))

log_msg("Generating predictions on test.csv...")
pred_test <- ifelse(predict(model, newdata = test_fe, type = "response") > 0.5, 1, 0)

out <- data.frame(
  PassengerId = test$PassengerId,
  Survived = pred_test
)
write_csv(out, "r_submission.csv")

log_msg("Saved predictions to: r_submission.csv")
log_msg("Note: test.csv has no 'Survived' column; skipping accuracy check as instructed.")
